Multimesh splitting

Norodix opened the pull request:
# Multimesh splitting is implemented for the plugin.

This feature is useful for large multimeshes that are created with the plugin, allowing culling to take place in batches.
The feature was tested on the showcase scene included in the plugin.

## Usage
A new checkbox is shown in the inspector of scatter nodes. 
The number of chunks can be specified for each axis with the Split N count parameter.
When the split multimesh checkbox is checked the chunks are created and saved under the scatter node and the original multimesh instance created by the plugin is hidden, but not deleted.
When the split multimesh checkbox is unchecked the split multimesh instances are deleted and the original multimesh instances are set visible.

HungryProton commented:
Thanks for the PR, seems to work well!

After trying it on an actual scene, I think there's only two things I would change to make it easier to use:
- Merge the `Split X, Split Y, Split Z` in a single Vector3 instead of three separate properties
- Move the root of the new multimeshes as a child of the ScatterItem (instead of the Scatter node itself)

The first point is to avoid cluttering the Scatter inspector too much.
The second point is because of this:  

![image](https://user-images.githubusercontent.com/52043844/184701615-a4094a3c-327d-494d-bdfd-0a02eab93f55.png)

Once the feature is enabled, it's hard to know which split Multimesh matches which ScatterItem. I think it makes more sense to hide them under their ScatterItem.

Also, if you have nested Scatter nodes like in the showcase scene, they also get split even if the option isn't enabled for them, and the splits are added to the parent, which doesn't seem right.

Other than that, the PR is good! If you can make these changes, it would be great, if you can't, I can merge it to a develop branch and do it myself, as you prefer.

Norodix commented:
I will try these changes.

Regarding the placement of the container under the scatter node itself. I did not know whether it would cause any problems to have multiple multimesh instances under the scatter item. Maybe the plugin gets a reference to it somewhere by finding a multimesh among its children? But if you say it doesnt then sure, I can move it under them no problem! Would be nicer too.

Regarding the recursive addition of split multimeshes to other scatter nodes. I noticed this behaviour and it took some extra effort to make it work correctly as well. It did not occur to me that it would be a problem, but you are right. It makes more sense to split only the ones that are explicitly selected. 

Vector3 for the split coutns works, but the only problem with it is its not very nice because you cant specify that a vector3 should only have ints as parameters. Are you sure that is a better approach? It has pros and cons, you should decide and ill make the change with the others. 

One more thing that I accidentally added to this pull request. (didnt know github adds all future commits to the branch to the PR) The disable autoupdate option. Its nice for larger meshes to be able to edit the path without it regenerating the mmis all the time. Do you agree?

On this same note, currently the way it works is after editing the scatter and getting it to how you like it you split it up to use it in-game. But with the split it isn't really possible to edit anymore. You need to turn off the split and edit then reenable it. Do you think it would be better to do this automatically for every update?

Norodix commented:
Also I have a very simplified LOD implementation to this same feature, I will create a PR once this one is sorted out!

Norodix commented:
What do you think of a submenu like this one?

![image](https://user-images.githubusercontent.com/36245389/184710447-c340f2bf-f128-4e64-888f-a65a535efab9.png)

HungryProton commented:
If you put the multimeshes like this, it shouldn't be a problem
![image](https://user-images.githubusercontent.com/52043844/184727511-87ff4953-1341-488f-8d30-fa396184d438.png)

Yes, I forgot the Vector3 doesn't allow for integers only, looks like I'm getting too used to Godot 4. Sub menus are better then. (I should probably take some time to sort the other properties too.)

For the auto update, you can set the preview count to 0, and it should behave the same way. This property limits how many points are handled no matter how the stack is set up to avoid performance issues. The feature is disabled by default (it's set to a negative value). 

For the split, there are a few options:
+ Do it every update
+ Ignore the splits completely in the editor and create them it in game only, in the _ready function
+ Add an extra button to manually regenerate the splits when necessary
+ Auto generate them when the node is deselected (or a few seconds after the last update)
 
Not sure which one is best.

Norodix commented:
I changed the tree structure, limited the split to the selected scatter and added a section in the editor.

One thing that I had a problem with was automatic updates. When the new split multimesh was added to the scatter item it triggered an update. Which of course refreshed the split multimesh, which triggered an update.....

So I added just added the refresh thing to the rebuild button. Of course disabling and re-enabling the split still works.

HungryProton commented:
Now merged, thanks again for the PR!
I've added an extra commit to regenerate the splits automatically when the user stops moving the curve, (and fixed the infinite loop issue at the same time), might need more testing just in case but it should be ok!
